url8d1df200a8_gilleain_signatures_tgz-pJ8-signature_chemistry_LargeMoleculeTestJ8/src/signature/display/DisplayableColoredTree.java

```diff
@@ -0,0 +1,275 @@
+        
+        public int depth;
+        
+        public int color;
+        
+        public List<DrawNode> children = new ArrayList<DrawNode>();
+        
+        public String label;
+        
+        public DrawNode parent;
+        
+        public String edgeLabel;
+        
+        public DrawNode(String label, DrawNode parent, int d, int color) {
+            this.label = label;
+            this.parent = parent;
+            this.depth = d;
+            this.color = color;
+            edgeLabel = "";
+        }
+        
+        public DrawNode(
+            String label, DrawNode parent, int d, int color, String edgeLabel) {
+            this(label, parent, d, color);
+            this.edgeLabel = edgeLabel;
+        }
+        
+        public int countLeaves() {
+            if (this.isLeaf()) {
+                return 1;
+            } else {
+                int c = 0;
+                for (DrawNode child : this.children) {
+                    c += child.countLeaves();
+                }
+                return c;
+            }
+        }
+        
+        public boolean isLeaf() {
+            return this.children.size() == 0;
+        }
+        
+        public String toString() {
+            StringBuffer buffer = new StringBuffer();
+            this.toString(buffer);
+            return buffer.toString();
+        }
+        
+        private void toString(StringBuffer buffer) {
+            buffer.append(label);
+            buffer.append('[').append(x).append(',').append(y).append(']');
+            for (DrawNode child : this.children) {
+                child.toString(buffer);
+            }
+        }
+    }
+    
+    public DisplayableColoredTree(int width, int height) {
+        this.root = null;
+        this.width = width;
+        this.height = height;
+        this.drawKey = false;
+    }
+    
+    public DisplayableColoredTree(ColoredTree tree, int width, int height) {
+        this.width = width;
+        this.height = height;
+        makeFromColoredTree(tree);
+        colorMap = makeColorMap(tree.numberOfColors());
+        this.drawKey = false;
+    }
+    
+    public void setDrawKey(boolean drawKey) {
+        this.drawKey = drawKey;
+    }
+    
+    private Map<Integer, Color> makeColorMap(int max) {
+        Map<Integer, Color> colorMap = new HashMap<Integer, Color>();
+        for (int i = 0; i <= max; i++) {
+            colorMap.put(i, colourRamp(i, 0, max));
+        }
+        return colorMap;
+    }
+    
+    public void makeFromColoredTree(ColoredTree tree) {
+        this.root = treeToTree(tree);
+        this.maxDepth = tree.getHeight();
+        this.colorMap = makeColorMap(tree.numberOfColors());
+    }
+    
+    private DrawNode treeToTree(ColoredTree tree) {
+        return treeToTree(tree.getRoot(), null);
+    }
+    
+    private DrawNode treeToTree(ColoredTree.Node treeNode, DrawNode parent) {
+        DrawNode node;
+        if (treeNode.edgeLabel == "") {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.